File: testing/fixtures/api/test-api-endpoint.template.js
// Template for API endpoint testing
// Copy this template and customize for your API endpoints

const request = require('supertest');

describe('API Endpoint Tests', () => {
  let app;
  let server;

  beforeAll(async () => {
    // Initialize your Express/Fastify app
    // app = require('./path/to/your/app');
    // server = app.listen();
  });

  afterAll(async () => {
    // Cleanup
    if (server) {
      await server.close();
    }
  });

  describe('GET /api/endpoint', () => {
    it('should return 200 OK for valid request', async () => {
      const response = await request(app)
        .get('/api/endpoint')
        .expect(200);

      expect(response.body).toHaveProperty('data');
      expect(response.body.data).toBeDefined();
    });

    it('should return 400 for invalid query parameters', async () => {
      const response = await request(app)
        .get('/api/endpoint')
        .query({ invalid: 'param' })
        .expect(400);

      expect(response.body).toHaveProperty('error');
    });

    it('should return 404 for non-existent resource', async () => {
      await request(app)
        .get('/api/endpoint/nonexistent')
        .expect(404);
    });
  });

  describe('POST /api/endpoint', () => {
    it('should create resource with valid data', async () => {
      const payload = {
        name: 'Test Resource',
        description: 'Test Description'
      };

      const response = await request(app)
        .post('/api/endpoint')
        .send(payload)
        .expect(201);

      expect(response.body).toHaveProperty('id');
      expect(response.body.name).toBe(payload.name);
    });

    it('should return 422 for invalid payload', async () => {
      const invalidPayload = {
        // Missing required fields
      };

      const response = await request(app)
        .post('/api/endpoint')
        .send(invalidPayload)
        .expect(422);

      expect(response.body).toHaveProperty('errors');
    });

    it('should return 401 for unauthorized request', async () => {
      await request(app)
        .post('/api/endpoint')
        .send({ name: 'Test' })
        .expect(401);
    });
  });

  describe('PUT /api/endpoint/:id', () => {
    it('should update resource with valid data', async () => {
      const resourceId = 'test-id-123';
      const updatePayload = {
        name: 'Updated Name'
      };

      const response = await request(app)
        .put(`/api/endpoint/${resourceId}`)
        .send(updatePayload)
        .expect(200);

      expect(response.body.name).toBe(updatePayload.name);
    });
  });

  describe('DELETE /api/endpoint/:id', () => {
    it('should delete resource successfully', async () => {
      const resourceId = 'test-id-123';

      await request(app)
        .delete(`/api/endpoint/${resourceId}`)
        .expect(204);
    });
  });

  describe('Health Check', () => {
    it('GET /health should return service status', async () => {
      const response = await request(app)
        .get('/health')
        .expect(200);

      expect(response.body).toHaveProperty('status', 'healthy');
      expect(response.body).toHaveProperty('timestamp');
    });

    it('GET /ready should return readiness status', async () => {
      const response = await request(app)
        .get('/ready')
        .expect(200);

      expect(response.body).toHaveProperty('ready', true);
      expect(response.body).toHaveProperty('checks');
    });
  });
});

File: testing/fixtures/api/test-api-endpoint.template.py
"""
Template for API endpoint testing in Python
Copy this template and customize for your API endpoints
"""
import pytest
from httpx import AsyncClient
from fastapi.testclient import TestClient

# If using FastAPI
# from your_app import app

# If using Flask
# from your_app import create_app


class TestAPIEndpoints:
    """Test API endpoints"""

    @pytest.fixture
    def client(self):
        """Setup test client"""
        # For FastAPI
        # return TestClient(app)

        # For Flask
        # app = create_app('testing')
        # return app.test_client()
        pass

    def test_get_endpoint_success(self, client):
        """Test GET endpoint returns 200 OK"""
        response = client.get('/api/endpoint')
        assert response.status_code == 200
        assert 'data' in response.json()

    def test_get_endpoint_with_params(self, client):
        """Test GET endpoint with query parameters"""
        response = client.get('/api/endpoint?param=value')
        assert response.status_code == 200
        data = response.json()
        assert data is not None

    def test_get_endpoint_not_found(self, client):
        """Test GET endpoint returns 404 for non-existent resource"""
        response = client.get('/api/endpoint/nonexistent')
        assert response.status_code == 404

    def test_post_endpoint_create_resource(self, client):
        """Test POST endpoint creates resource"""
        payload = {
            'name': 'Test Resource',
            'description': 'Test Description'
        }
        response = client.post('/api/endpoint', json=payload)
        assert response.status_code == 201
        data = response.json()
        assert 'id' in data
        assert data['name'] == payload['name']

    def test_post_endpoint_invalid_payload(self, client):
        """Test POST endpoint returns 422 for invalid payload"""
        invalid_payload = {}
        response = client.post('/api/endpoint', json=invalid_payload)
        assert response.status_code == 422
        assert 'detail' in response.json() or 'errors' in response.json()

    def test_post_endpoint_unauthorized(self, client):
        """Test POST endpoint returns 401 for unauthorized request"""
        response = client.post('/api/endpoint', json={'name': 'Test'})
        assert response.status_code == 401

    def test_put_endpoint_update_resource(self, client):
        """Test PUT endpoint updates resource"""
        resource_id = 'test-id-123'
        update_payload = {'name': 'Updated Name'}
        response = client.put(f'/api/endpoint/{resource_id}', json=update_payload)
        assert response.status_code == 200
        data = response.json()
        assert data['name'] == update_payload['name']

    def test_delete_endpoint_success(self, client):
        """Test DELETE endpoint removes resource"""
        resource_id = 'test-id-123'
        response = client.delete(f'/api/endpoint/{resource_id}')
        assert response.status_code == 204

    def test_health_check(self, client):
        """Test health check endpoint"""
        response = client.get('/health')
        assert response.status_code == 200
        data = response.json()
        assert data['status'] == 'healthy'
        assert 'timestamp' in data

    def test_readiness_check(self, client):
        """Test readiness check endpoint"""
        response = client.get('/ready')
        assert response.status_code == 200
        data = response.json()
        assert data['ready'] is True
        assert 'checks' in data


class TestAPIEndpointsAsync:
    """Test async API endpoints"""

    @pytest.fixture
    async def async_client(self):
        """Setup async test client"""
        # For FastAPI with async
        # async with AsyncClient(app=app, base_url="http://test") as client:
        #     yield client
        pass

    @pytest.mark.asyncio
    async def test_get_endpoint_async(self, async_client):
        """Test GET endpoint asynchronously"""
        response = await async_client.get('/api/endpoint')
        assert response.status_code == 200
        assert 'data' in response.json()

    @pytest.mark.asyncio
    async def test_post_endpoint_async(self, async_client):
        """Test POST endpoint asynchronously"""
        payload = {'name': 'Test Resource'}
        response = await async_client.post('/api/endpoint', json=payload)
        assert response.status_code == 201
        data = response.json()
        assert 'id' in data
